File: app/schemas/source.py
from datetime import datetime
from typing import Optional, Literal, Dict, Any

from pydantic import BaseModel, ConfigDict, model_validator


class SourceBase(BaseModel):
    name: str
    secret: Optional[str] = None  # Secret key (required when auth_type is "signature", optional when "none")
    auth_type: Literal["signature", "none"] = "none"  # 認證類型
    auth_config: Optional[Dict[str, Any]] = None  # 認證配置 (JSON)

    @model_validator(mode='after')
    def validate_secret_for_signature(self):
        """驗證當 auth_type 為 signature 時，secret 必須提供"""
        if self.auth_type == "signature" and not self.secret:
            raise ValueError("當認證類型為 'signature' 時，secret 為必填項目")
        return self


class SourceCreate(SourceBase):
    pass


class SourceUpdate(BaseModel):
    name: Optional[str] = None
    secret: Optional[str] = None
    auth_type: Optional[Literal["signature", "none"]] = None
    auth_config: Optional[Dict[str, Any]] = None


class SourceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str  # 改為 str 支援 ULID
    name: str
    auth_type: str
    auth_config: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: Optional[datetime] = None

File: frontend/src/types/source.ts
/**
 * 來源相關的 TypeScript 類型定義
 */

// 認證類型枚舉
export type AuthType = 'signature' | 'none'

// 來源基本資訊
export interface SourceBase {
  name: string
  secret: string
  auth_type: AuthType
  auth_config?: Record<string, any>
}

// 創建來源請求
export interface SourceCreate extends SourceBase {}

// 來源響應
export interface SourceResponse {
  id: string
  name: string
  auth_type: AuthType
  auth_config?: Record<string, any>
  created_at: string
  updated_at?: string
}

// 來源列表響應包裝器
export interface SourceListResponse {
  items: SourceResponse[]
  total: number
  skip: number
  limit: number
}

// 來源服務錯誤
export interface SourceError {
  message: string
  code?: string
  field?: string
}

// API 響應包裝器
export interface ApiResponse<T> {
  data?: T
  error?: SourceError
  loading: boolean
}

// 分頁參數
export interface PaginationParams {
  skip?: number
  limit?: number
}

// 來源篩選參數
export interface SourceFilterParams extends PaginationParams {
  name?: string
  auth_type?: AuthType
}

// 來源統計數據
export interface SourceStats {
  source_id: string
  webhook_count: number
  topic_count: number
  last_activity?: string
}

// 更新來源請求
export interface SourceUpdate {
  name?: string
  secret?: string
  auth_type?: AuthType
  auth_config?: Record<string, any>
}

// 來源表單驗證錯誤
export interface SourceFormErrors {
  name?: string
  secret?: string
  auth_type?: string
  auth_config?: string
}
